plugins/algoltek-aux/fu-algoltek-aux-device.c

```diff
@@ -0,0 +1,517 @@
+	g_autoptr(GInputStream) stream_isp = NULL;
+	g_autoptr(GInputStream) stream_payload = NULL;
+
+	/* progress */
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 18, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 2, "isp");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 80, "payload");
+
+	/* prepare hardware? */
+	if (!fu_algoltek_aux_device_en(self, error))
+		return FALSE;
+	if (!fu_algoltek_aux_device_rst(self, error))
+		return FALSE;
+	fu_device_sleep(FU_DEVICE(self), 500);
+	if (!fu_algoltek_aux_device_wrr(self, 0x80AD, 0, error))
+		return FALSE;
+	if (!fu_algoltek_aux_device_wrr(self, 0x80C0, 0, error))
+		return FALSE;
+	if (!fu_algoltek_aux_device_wrr(self, 0x80C9, 0, error))
+		return FALSE;
+	if (!fu_algoltek_aux_device_wrr(self, 0x80D1, 0, error))
+		return FALSE;
+	if (!fu_algoltek_aux_device_wrr(self, 0x80D9, 0, error))
+		return FALSE;
+	if (!fu_algoltek_aux_device_wrr(self, 0x80E1, 0, error))
+		return FALSE;
+	if (!fu_algoltek_aux_device_wrr(self, 0x80E9, 0, error))
+		return FALSE;
+	fu_device_sleep(FU_DEVICE(self), 20);
+	fu_progress_step_done(progress);
+
+	/* write ISP image */
+	stream_isp = fu_firmware_get_image_by_id_stream(firmware, "isp", error);
+	if (stream_isp == NULL)
+		return FALSE;
+	if (!fu_algoltek_aux_device_isp(self,
+					stream_isp,
+					&wcrc,
+					fu_progress_get_child(progress),
+					error))
+		return FALSE;
+
+	fu_device_sleep(FU_DEVICE(self), 2000);
+	if (!fu_algoltek_aux_device_bot(self, 0x6000, error))
+		return FALSE;
+	fu_device_sleep(FU_DEVICE(self), 2000);
+	if (!fu_algoltek_aux_device_ers(self, error))
+		return FALSE;
+	fu_device_sleep(FU_DEVICE(self), 5000);
+	fu_progress_step_done(progress);
+
+	/* write payload image */
+	stream_payload =
+	    fu_firmware_get_image_by_id_stream(firmware, FU_FIRMWARE_ID_PAYLOAD, error);
+	if (stream_payload == NULL)
+		return FALSE;
+	if (!fu_algoltek_aux_device_wrf(self,
+					stream_payload,
+					&wcrc,
+					fu_progress_get_child(progress),
+					error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_aux_device_setup(FuDevice *device, GError **error)
+{
+	FuAlgoltekAuxDevice *self = FU_ALGOLTEK_AUX_DEVICE(device);
+	g_autofree gchar *product = NULL;
+	g_autofree gchar *version_str = NULL;
+	g_autoptr(GByteArray) version_data = NULL;
+
+	/* FuAuxDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_algoltek_aux_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	/* get current version */
+	version_data = fu_algoltek_aux_device_rdv(self, error);
+	if (version_data == NULL)
+		return FALSE;
+	version_str = fu_strsafe((const gchar *)version_data->data, version_data->len);
+	fu_device_set_version(FU_DEVICE(self), version_str);
+
+	/* build something unique as a GUID */
+	fu_device_add_instance_strup(device, "VEN", "25A4");
```

Comment:
does this also happen to come in `fu_algoltek_aux_device_rdv` IE through `version_data`?

Reply:
@YiJaneWu any ideas?

Reply:
Yes, while reading the version number, I also read the product category.
I take the product category as a GUID.

Reply:
Okay that should be better than hardcoding.

Reply:
@superm1 the `FuStructAlgoltekAuxProductIdentity` header is literally `header: u64le == 0x4B45544C4F474C41, // 'A' 'L' 'G' 'O' 'L' 'T' 'E' 'K'` -- I think it's okay to hardcode :)